src/legacy/server/saved_objects/service/saved_objects_client.ts

```diff
@@ -0,0 +1,307 @@
+
+type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
+
+export interface BaseOptions {
+  /** Specify the namespace for this operation */
+  namespace?: string;
+}
+
+export interface CreateOptions extends BaseOptions {
+  /** (not recommended) Specify an id for the document */
+  id?: string;
+  /** Overwrite existing documents (defaults to false) */
+  overwrite?: boolean;
+  migrationVersion?: MigrationVersion;
+  references?: SavedObjectReference[];
+}
+
+export interface BulkCreateObject<T extends SavedObjectAttributes = any> {
+  id?: string;
+  type: string;
+  attributes: T;
+  references?: SavedObjectReference[];
+  migrationVersion?: MigrationVersion;
+}
+
+export interface BulkResponse<T extends SavedObjectAttributes = any> {
+  saved_objects: Array<SavedObject<T>>;
+}
+
+export interface FindOptions extends BaseOptions {
+  type?: string | string[];
+  page?: number;
+  perPage?: number;
+  sortField?: string;
+  sortOrder?: string;
+  fields?: string[];
+  search?: string;
+  /** see Elasticsearch Simple Query String Query field argument for more information */
+  searchFields?: string[];
+  hasReference?: { type: string; id: string };
+  defaultSearchOperator?: 'AND' | 'OR';
+}
+
+export interface FindResponse<T extends SavedObjectAttributes = any> {
+  saved_objects: Array<SavedObject<T>>;
+  total: number;
+  per_page: number;
+  page: number;
+}
+
+export interface UpdateOptions extends BaseOptions {
+  /** Ensures version matches that of persisted object */
+  version?: string;
+  references?: SavedObjectReference[];
+}
+
+export interface BulkGetObject {
+  id: string;
+  type: string;
+  /** SavedObject fields to include in the response */
+  fields?: string[];
+}
+
+export interface BulkResponse<T extends SavedObjectAttributes = any> {
+  saved_objects: Array<SavedObject<T>>;
+}
+
+export interface UpdateResponse<T extends SavedObjectAttributes = any>
+  extends Omit<SavedObject<T>, 'attributes'> {
+  attributes: Partial<T>;
+}
+
+/**
+ * A dictionary of saved object type -> version used to determine
+ * what migrations need to be applied to a saved object.
+ */
+export interface MigrationVersion {
+  [pluginName: string]: string;
+}
+
+export interface SavedObjectAttributes {
+  [key: string]: SavedObjectAttributes | string | number | boolean | null;
+}
+
+export interface VisualizationAttributes extends SavedObjectAttributes {
+  visState: string;
+}
+
+export interface SavedObject<T extends SavedObjectAttributes = any> {
+  id: string;
+  type: string;
+  version?: string;
+  updated_at?: string;
+  error?: {
+    message: string;
+    statusCode: number;
+  };
+  attributes: T;
+  references: SavedObjectReference[];
+  migrationVersion?: MigrationVersion;
+}
+
+/**
+ * A reference to another saved object.
+ */
+export interface SavedObjectReference {
+  name: string;
+  type: string;
+  id: string;
+}
+
+export type SavedObjectsClientContract = Pick<SavedObjectsClient, keyof SavedObjectsClient>;
```

Comment:
I decided to use a similar pattern to `PublicMethodsOf` from this discussion, the pro's and cons given there are also relevant to this PR: https://github.com/elastic/kibana/issues/33396 

This type differs from `PublicMethodsOf` in that it also includes public properties like `SavedObjectsClient.errors` which in this case is an important part of the `SavedObjectsClient` API